### PR TITLE
Week6 : JS Advanced Concepts

### PR DESCRIPTION
**Pull request by rajarshi1**

### CODE REVIEW

File: Week-6/ES6Module/calculator.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// Define the add function
+export function add(a, b) {
+    return a + b;
+}
+
+// Define the subtract function
+export function subtract(a, b) {
```
Comment (sraghudatta94) [inline Week-6/ES6Module/calculator.js:7]: Consider including a default export in your modules if there is a primary functionality that is expected to be used frequently.

File: Week-6/maps&set/script.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
+// WeakMap for storing product views
+const productViews = new WeakMap();
+
+// WeakSet for storing product IDs that are in the user's cart
+const cartItems = new WeakSet();
+
+// Function to increment product views
+function incrementProductViews(productId) {
```
Comment (sraghudatta94) [inline Week-6/maps&set/script.js:8]: In the current implementation, productRef is a new object created every time `incrementProductViews `or `addToCart `is called, which means it won't match the previous references stored in `productViews `or `cartItems`. Instead, these should be persistent references that correspond to product IDs.

Sharing for learning!

File: Week-6/tictactoe/tictactoe.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  // Create click-event handlers for each game board button
+  const buttons = getGameBoardButtons();
+  for (let button of buttons) {
+    button.addEventListener("click", function () {
+      boardButtonClicked(button);
+    });
+  }
+
+  // Clear the board
+  newGame();
```
Comment (sraghudatta94) [inline Week-6/tictactoe/tictactoe.js:27]: There's a risk of a memory leak with computerMoveTimeout if newGame is called before the timeout is completed. Good to see it being cleared on new game starts, just make sure it's consistently handled on all relevant events.

Centralize timeout management
```
function setComputerMoveTimeout() {
  clearComputerMoveTimeout(); // Clear any existing timeout
  computerMoveTimeout = setTimeout(makeComputerMove, 1000);
}
```

```
function clearComputerMoveTimeout() {
  if (computerMoveTimeout) {
    clearTimeout(computerMoveTimeout);
    computerMoveTimeout = 0;
  }
}
```


Update within switchTurn function
```
function switchTurn() {
  // ... existing code ...
  if (!playerTurn) {
    setComputerMoveTimeout();
  }
  // ... existing code ...
}
```

Update within newGame function
```
function newGame() {
  clearComputerMoveTimeout();
  // ... remaining logic for newGame ...
}
```

File: Week-6/windowsObject/index.html [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <title>Window Object Properties</title>
+    <script src="script.js" defer></script>
+</head>
+<body onload="displayWindowProperties()">
```
Comment (sraghudatta94) [inline Week-6/windowsObject/index.html:8]: Since `window.onload` is used at the bottom of the script, the `onload `attribute in the `body `tag is redundant and could be removed.

**Review by sraghudatta94 — APPROVED** [review]

**Comment by sraghudatta94** [comment]
Overall, you are doing an excellent job of applying advanced JavaScript concepts in a practical manner. Your code is organized and follows many of the best practices. Keep exploring more ways to ensure code robustness, such as writing tests and handling edge cases. Your progress is evident, and with a focus on continuous improvement, your skills will become even sharper. Great work!
